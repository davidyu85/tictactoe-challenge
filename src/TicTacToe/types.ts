export interface TicTacToeProps {
  wildMode?: boolean;
  computerIsPlayer2?: boolean;
}

export type ChessPiece = 'X' | 'O';

export type TicTacToeGameState = (ChessPiece | null)[];

export type ChessAccessMap = { computer: ChessPiece[]; player: ChessPiece[] };

export type GameplayActions =
  | {
      type: 'place-chess';
      cellPos: number;
      chess: ChessPiece;
    }
  | {
      type: 'computer-place-chess';
      wildMode: boolean;
    };

export interface GameplayStates {
  gameState: TicTacToeGameState;
  isPlayerTwoTurn: boolean;
  isPlayerWins: boolean;
  isDraw: boolean;
}
