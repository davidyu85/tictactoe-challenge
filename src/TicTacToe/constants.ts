export const SINGLE_PLAYER_STRING = 'Player VS Computer';

export const TWO_PLAYER_TURN_STRING = {
  1: 'Player 1 turn',
  2: 'Player 2 turn',
};

export const SINGLE_PLAYER_WIN_STRING = {
  1: 'Player wins!',
  2: 'Computer wins!',
};

export const TWO_PLAYER_WIN_STRING = {
  1: 'Player 1 wins!',
  2: 'Player 2 wins!',
};

export const DRAW_STRING = 'Draw!';
