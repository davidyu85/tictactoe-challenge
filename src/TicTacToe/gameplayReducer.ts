import { CHESS_MAP } from './constants';
import { GameplayStates, GameplayActions, ChessPiece } from './types';
import checkEndGame from './utils/checkEndGame';
import miniMax from './utils/miniMax';

export const initState: GameplayStates = {
  gameBoard: [...Array(9).fill(null)],
  isPlayerTwoTurn: false,
  isPlayerWins: false,
  isDraw: false,
};

const gameplayReducer = (
  state: GameplayStates,
  action: GameplayActions
): GameplayStates => {
  const { gameBoard, isPlayerTwoTurn } = state;

  switch (action.type) {
    case 'place-chess': {
      const board = [...gameBoard];
      board[action.cellPos] = action.chess;

      const endGameState = checkEndGame(board, isPlayerTwoTurn);

      return {
        ...state,
        gameBoard: board,
        ...endGameState,
      };
    }

    case 'computer-place-chess': {
      const board = [...gameBoard];
      const chessMap = CHESS_MAP[action.wildMode ? 'wild' : 'standard'];
      let bestMove = { score: -Infinity, pos: -1, chess: chessMap.computer[0] };

      chessMap.computer.forEach((chess: ChessPiece) => {
        for (let pos = 0; pos < board.length; pos++) {
          if (board[pos] !== null) continue;

          board[pos] = chess;
          const score = miniMax(chessMap, board, false, -Infinity, Infinity);
          board[pos] = null;

          if (score > bestMove.score) {
            bestMove = { score, chess, pos };
          }
        }
      });

      board[bestMove.pos] = bestMove.chess;

      const endGameState = checkEndGame(board, isPlayerTwoTurn);

      return {
        ...state,
        gameBoard: board,
        ...endGameState,
      };
    }
  }
};

export default gameplayReducer;
