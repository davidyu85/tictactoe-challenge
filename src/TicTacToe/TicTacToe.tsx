import { ReactElement, useEffect, useMemo, useReducer, useRef } from 'react';
import { BoardCell, GameBoard } from './Board/Board';
import gameplayReducer, { initState } from './gameplayReducer';
import { SINGLE_PLAYER_STRING, TWO_PLAYER_TURN_STRING } from './constants';
import ChessSelect from './ChessSelect/ChessSelect';
import GameEnds from './GameEnds/GameEnds';
import { ChessPiece, Players } from './types';
import singlePlayerChessSetup from './utils/singlePlayerChessSetup';

export interface TicTacToeProps {
  /** Toggle between standard gameplay or wild variant */
  wildMode?: boolean;
  /** Toggle the second player to be computer or a human */
  computerIsPlayer?: Players;
}

/**
 * A proof of concept of Tic Tac Toe game, supporting
 * - standard 3 X 3 grid Tic Tac Toe gameplay
 * - wild variant, allowing players to play both X and O
 * - single player game play with a computer player
 *
 * @param {TicTacToeProps} props
 * @returns {ReactElement}
 */
const TicTacToe = ({
  wildMode,
  computerIsPlayer,
}: TicTacToeProps): ReactElement => {
  const [states, dispatch] = useReducer(gameplayReducer, initState);
  const { gameBoard, isPlayerTwoTurn, isPlayerWins, isDraw } = states;
  const chessSelect = useRef<HTMLSelectElement>(null);

  const handleClickToPlaceChess = (cellPos: number) => (): void => {
    dispatch({
      type: 'place-chess',
      cellPos,
      chess: chessSelect.current?.value as ChessPiece,
    });
  };

  const singlePlayer = useMemo(
    () => singlePlayerChessSetup(isPlayerTwoTurn, wildMode, computerIsPlayer),
    [isPlayerTwoTurn, wildMode, computerIsPlayer]
  );

  const disableBoard: boolean | undefined =
    isPlayerWins || isDraw || singlePlayer?.isComputerPlayerTurn;

  useEffect(() => {
    if (singlePlayer?.isComputerPlayerTurn) {
      dispatch({
        type: 'computer-place-chess',
        chessMap: singlePlayer.chessMap,
      });
    }
  }, [singlePlayer]);

  return (
    <>
      <GameBoard>
        {gameBoard.map((chess, i) => (
          <BoardCell
            key={`cell-${i}`}
            chess={chess}
            onClickOnce={handleClickToPlaceChess(i)}
            disabled={disableBoard}
          />
        ))}
      </GameBoard>

      <label>
        {computerIsPlayer
          ? SINGLE_PLAYER_STRING[computerIsPlayer]
          : TWO_PLAYER_TURN_STRING[isPlayerTwoTurn ? 2 : 1]}
      </label>

      {!isDraw && !isPlayerWins && (
        <ChessSelect
          ref={chessSelect}
          wildMode={wildMode}
          isPlayerTwoTurn={isPlayerTwoTurn}
        />
      )}

      <GameEnds
        computerIsPlayer={computerIsPlayer}
        isPlayerTwoTurn={isPlayerTwoTurn}
        isPlayerWins={isPlayerWins}
        isDraw={isDraw}
      />
    </>
  );
};

export default TicTacToe;
