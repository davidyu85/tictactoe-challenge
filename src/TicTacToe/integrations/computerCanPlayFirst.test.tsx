import { describe, it, expect, afterEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import TicTacToe from '../TicTacToe';
import { SINGLE_PLAYER_WIN_STRING } from '../constants';
import { renderGameStateSnapShot } from '../../../vitest.setup';

describe('Computer can play first', () => {
  afterEach(() => {
    renderGameStateSnapShot();
  });

  it('can finish a game in standard mode', () => {
    render(<TicTacToe computerIsPlayer={1} />);
    expect(screen.getByRole('button', { name: 'X' })).toBeInTheDocument();

    fireEvent.click(screen.getAllByRole('button')[1]);
    fireEvent.click(screen.getAllByRole('button')[6]);
    fireEvent.click(screen.getAllByRole('button')[8]);

    expect(screen.getByText(SINGLE_PLAYER_WIN_STRING[2])).toBeVisible();
  });

  it('can finish a game in wild mode', () => {
    render(<TicTacToe computerIsPlayer={1} wildMode />);
    expect(screen.getByRole('button', { name: 'X' })).toBeInTheDocument();

    fireEvent.change(screen.getByDisplayValue('X'), { target: { value: 'O' } });
    fireEvent.click(screen.getAllByRole('button')[0]);
    fireEvent.click(screen.getAllByRole('button')[2]);

    expect(screen.getByText(SINGLE_PLAYER_WIN_STRING[2])).toBeVisible();
  });
});
