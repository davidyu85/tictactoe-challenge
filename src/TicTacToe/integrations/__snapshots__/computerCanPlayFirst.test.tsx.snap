// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Computer can play first > can finish a game in standard mode 1`] = `
"
 X O  
 X X X
 O   O
"
`;

exports[`Computer can play first > can finish a game in wild mode 1`] = `
"
 O O O
   X  
     O
"
`;
